tests/autobahn/test_autobahn.py

```diff
@@ -0,0 +1,163 @@
+        subprocess.check_call(
+            ["virtualenv", "-p", "python2.7", "tests/autobahn/autobahntestsuite-env"]
+        )
+        subprocess.check_call(
+            [
+                "tests/autobahn/autobahntestsuite-env/bin/pip",
+                "install",
+                "autobahntestsuite>=0.8.0",
+            ]
+        )
+
+
+def wait_for_server(host: str, port: int) -> None:
+    while True:
+        sock = socket.socket()
+        try:
+            sock.connect((HOST, PORT))
+        except OSError as sock_err:
+            if sock_err.errno == errno.ECONNREFUSED:
+                time.sleep(0.01)
+            else:
+                raise
+        else:
+            return
+        finally:
+            sock.close()
+
+
+def get_failed_tests(report_path: str, name: str) -> List[Dict[str, Any]]:
+    with open(Path(f"{report_path}/index.json")) as f:
+        result_summary = json.load(f)[name]
+    failed_messages = []
+    PASS = {"OK", "INFORMATIONAL"}
+    for results in result_summary.values():
+        if results["behavior"] not in PASS or results["behaviorClose"] not in PASS:
+            with open(Path(f"{report_path}/{results['reportfile']}")) as f:
+                report = json.load(f)
+                failed_messages.append(
+                    {
+                        "case": report["case"],
+                        "description": report["description"],
+                        "expectation": report["expectation"],
+                        "expected": report["expected"],
+                        "received": report["received"],
+                    }
+                )
+    return failed_messages
+
+
+def test_client() -> None:
+    print("Starting autobahntestsuite server")
+    server = subprocess.Popen(
+        [
+            "tests/autobahn/autobahntestsuite-env/bin/wstest",
+            "-m",
+            "fuzzingserver",
+            "-s",
+            "tests/autobahn/client/fuzzingserver.json",
+        ]
+    )
+    print("Waiting for server to start")
+    wait_for_server(HOST, PORT)
+    try:
+        print("Running wsproto test client")
+        client = subprocess.Popen(
+            [sys.executable] + ["tests/autobahn/client/client.py"]
+        )
+        client.wait()
+    finally:
+        print("Stopping server")
+        client.terminate()
+        client.wait()
+        server.terminate()
+        server.wait()
+
+    failed_messages = get_failed_tests("tests/autobahn/reports/clients", "aiohttp")
+
+    if failed_messages:
+        pytest.xfail(
+            "\n".join(
+                [
+                    f"case: {msg['case']}"
+                    f"\ndescription: {msg['description']}"
+                    f"\nexpectation: {msg['expectation']}"
+                    f"\nexpected: {msg['expected']}"
+                    f"\nreceived: {msg['received']}"
+                    for msg in failed_messages
+                ]
+            )
+        )
```

Comment:
It's best to use an actual `assert` statement:
```suggestion
    assert not failed_messages, "\n".join(
        f"case: {msg['case']}"
        f"\ndescription: {msg['description']}"
        f"\nexpectation: {msg['expectation']}"
        f"\nexpected: {msg['expected']}"
        f"\nreceived: {msg['received']}"
        for msg in failed_messages
    )
```